### PR TITLE
NOTIF-138 Replace R2DBC with Hibernate Reactive

### PR DESCRIPTION
**Pull request by gwenneg**
Don't try to build the branch yet, it won't work.

I'm still working on some issues, especially this one:
```
2021-03-09 22:41:56,286 WARN  [org.hib.eng.jdb.spi.SqlExceptionHelper] (vert.x-eventloop-thread-6) SQL Error: 0, SQLState: null
2021-03-09 22:41:56,287 ERROR [org.hib.eng.jdb.spi.SqlExceptionHelper] (vert.x-eventloop-thread-6) Not using JDBC
2021-03-09 22:41:56,290 ERROR [org.hib.rea.errors] (vert.x-eventloop-thread-6) failed to execute statement [select endpointen0_.id as id1_7_, endpointen0_.created as created2_7_, endpointen0_.updated as updated3_7_, endpointen0_.account_id as account_4_7_, endpointen0_.description as descript5_7_, endpointen0_.enabled as enabled6_7_, endpointen0_.endpoint_type as endpoint7_7_, endpointen0_.name as name8_7_ from endpoints endpointen0_ where endpointen0_.id=$1 and endpointen0_.account_id=$2]
2021-03-09 22:41:56,292 ERROR [org.hib.rea.errors] (vert.x-eventloop-thread-6) could not execute query: java.util.concurrent.CompletionException: org.hibernate.exception.GenericJDBCException: Unable to acquire JDBC Connection
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
        at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(CompletableFuture.java:670)
        at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:658)
        at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:2094)
        at java.base/java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:143)
        at com.ibm.asyncutil.util.StageSupport.voided(StageSupport.java:53)
        at com.ibm.asyncutil.iteration.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:209)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:169)
        at org.hibernate.reactive.util.impl.CompletionStages.loop(CompletionStages.java:186)
        at org.hibernate.reactive.loader.ReactiveLoaderBasedResultSetProcessor.reactiveInitializeEntitiesAndCollections(ReactiveLoaderBasedResultSetProcessor.java:141)
        at org.hibernate.reactive.loader.ReactiveLoaderBasedResultSetProcessor.reactiveExtractResults(ReactiveLoaderBasedResultSetProcessor.java:83)
        at org.hibernate.reactive.loader.hql.impl.ReactiveQueryLoader$1.reactiveExtractResults(ReactiveQueryLoader.java:72)
        at org.hibernate.reactive.loader.ReactiveLoader.reactiveProcessResultSet(ReactiveLoader.java:146)
        at org.hibernate.reactive.loader.ReactiveLoader.lambda$doReactiveQueryAndInitializeNonLazyCollections$0(ReactiveLoader.java:77)
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1072)
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073)
        at org.hibernate.reactive.pool.impl.Handlers.lambda$toCompletionStage$0(Handlers.java:26)
        at io.vertx.sqlclient.impl.SqlResultHandler.complete(SqlResultHandler.java:98)
        at io.vertx.sqlclient.impl.SqlResultHandler.handle(SqlResultHandler.java:87)
        at io.vertx.sqlclient.impl.SqlResultHandler.handle(SqlResultHandler.java:33)
        at io.vertx.sqlclient.impl.SocketConnectionBase.handleMessage(SocketConnectionBase.java:241)
        at io.vertx.sqlclient.impl.SocketConnectionBase.lambda$init$0(SocketConnectionBase.java:88)
        at io.vertx.core.net.impl.NetSocketImpl.lambda$new$2(NetSocketImpl.java:101)
        at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:237)
        at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:127)
        at io.vertx.core.net.impl.NetSocketImpl.handleMessage(NetSocketImpl.java:357)
        at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:366)
        at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:43)
        at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:229)
        at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:163)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
        at io.vertx.pgclient.impl.codec.PgEncoder.lambda$write$0(PgEncoder.java:78)
        at io.vertx.pgclient.impl.codec.PgCommandCodec.handleReadyForQuery(PgCommandCodec.java:138)
        at io.vertx.pgclient.impl.codec.PgDecoder.decodeReadyForQuery(PgDecoder.java:226)
        at io.vertx.pgclient.impl.codec.PgDecoder.channelRead(PgDecoder.java:86)
        at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:714)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:576)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:493)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.hibernate.exception.GenericJDBCException: Unable to acquire JDBC Connection
        at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:47)
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
        at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
        at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:111)
        at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.getPhysicalConnection(LogicalConnectionManagedImpl.java:138)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.connection(StatementPreparerImpl.java:50)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$5.doPrepare(StatementPreparerImpl.java:149)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl$StatementPreparationTemplate.prepareStatement(StatementPreparerImpl.java:176)
        at org.hibernate.engine.jdbc.internal.StatementPreparerImpl.prepareQueryStatement(StatementPreparerImpl.java:151)
        at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:2103)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2040)
        at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2018)
        at org.hibernate.loader.Loader.doQuery(Loader.java:948)
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:349)
        at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:319)
        at org.hibernate.loader.Loader.loadEntity(Loader.java:2421)
        at org.hibernate.loader.entity.AbstractEntityLoader.load(AbstractEntityLoader.java:72)
        at org.hibernate.loader.entity.EntityLoader.loadByUniqueKey(EntityLoader.java:148)
        at org.hibernate.loader.entity.EntityLoader.loadByUniqueKey(EntityLoader.java:144)
        at org.hibernate.persister.entity.AbstractEntityPersister.loadByUniqueKey(AbstractEntityPersister.java:2475)
        at org.hibernate.type.EntityType.loadByUniqueKey(EntityType.java:754)
        at org.hibernate.type.EntityType.resolve(EntityType.java:467)
        at org.hibernate.reactive.loader.ReactiveResultSetProcessor.lambda$initializeEntity$0(ReactiveResultSetProcessor.java:79)
        at org.hibernate.engine.internal.TwoPhaseLoad.initializeEntityEntryLoadedState(TwoPhaseLoad.java:248)
        at org.hibernate.reactive.loader.ReactiveResultSetProcessor.initializeEntity(ReactiveResultSetProcessor.java:72)
        at org.hibernate.reactive.loader.ReactiveLoaderBasedResultSetProcessor.lambda$reactiveInitializeEntitiesAndCollections$2(ReactiveLoaderBasedResultSetProcessor.java:143)
        at org.hibernate.reactive.util.impl.CompletionStages.lambda$loop$5(CompletionStages.java:169)
        at com.ibm.asyncutil.iteration.AsyncTrampoline.lambda$asyncWhile$1(AsyncTrampoline.java:209)
        at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.unroll(AsyncTrampoline.java:119)
        at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.trampoline(AsyncTrampoline.java:101)
        at com.ibm.asyncutil.iteration.AsyncTrampoline$TrampolineInternal.access$200(AsyncTrampoline.java:84)
        at com.ibm.asyncutil.iteration.AsyncTrampoline.asyncWhile(AsyncTrampoline.java:192)
        ... 50 more
Caused by: java.sql.SQLException: Not using JDBC
        at org.hibernate.reactive.provider.service.NoJdbcConnectionProvider.getConnection(NoJdbcConnectionProvider.java:25)
        at org.hibernate.internal.NonContextualJdbcConnectionAccess.obtainConnection(NonContextualJdbcConnectionAccess.java:38)
        at org.hibernate.resource.jdbc.internal.LogicalConnectionManagedImpl.acquireConnectionIfNeeded(LogicalConnectionManagedImpl.java:108)
        ... 78 more
```

I talked about it with Gavin, it may be an issue in Hibernate Reactive. Investigation in progress...

### CODE REVIEW

**Comment by gwenneg** [comment]
The failed build is normal, I'll push the new version of `EndpointResources` after I'm done fixing an issue involving this class.

File: pom.xml [hunk 1]
```diff
@@ -228,14 +223,19 @@
                         </goals>
                     </execution>
                 </executions>
-                <configuration>
```
Comment (pilhuhn) [inline pom.xml:231]: Why is this removed?
We did chose port 5007 on purpose as other processes already use 5005+5006

**Comment by pilhuhn** [comment]
Two general questions:
1) why do you use DTos instead of only the model classes?
2) Is mapstruct "native-safe"? While we don't immediately want to go native, we should not block doing so.

File: src/main/java/com/redhat/cloud/notifications/db/entities/ApplicationEntity.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+public class ApplicationEntity extends CreationUpdateTimestampedEntity {
+
+    @Id
+    @GeneratedValue
+    @NotNull
+    public UUID id;
+
+    @NotNull
+    @Size(max = 255)
+    @Pattern(regexp = "[a-z][a-z_0-9-]*")
```
Comment (pilhuhn) [inline src/main/java/com/redhat/cloud/notifications/db/entities/ApplicationEntity.java:30]: If we keep the separation Model / Entity : would that be needed at all, as the validations are already on data-intake via the model.

**Comment by gwenneg** [comment]
I worked on several projects in the past where the API and persistence (Hibernate-based) layers were tightly coupled because of a common model. While the common model way might be seducing at first, I have yet to see a project that doesn't suffer from this choice in terms of maintainability in the long term. On the other hand, I worked on many projects with separate API and persistence models and it was always easier to maintain.

I haven't checked yet Mapstruct native compatibilty, that's a good question. TBH, I'm not sure I want to keep Mapstruct at all in this PR because I had to write many specific mapping behaviors.

File: pom.xml (hunk 1 above)
Comment (gwenneg) [reply pom.xml]: Fixed.

File: src/main/java/com/redhat/cloud/notifications/routers/UserConfigService.java [hunk 3]
```diff
@@ -72,6 +72,10 @@
                     }
                 })));
 
+        /*
+         * (Un)subscriptions will be run sequentially because of the concatenation.
+         * This is mandatory because the Hibernate session must not be used concurrently.
+         */
```
Comment (josejulio) [inline src/main/java/com/redhat/cloud/notifications/routers/UserConfigService.java:78]: IIUC we should never use "merge" when dealing with the database?

Comment (gwenneg) [reply src/main/java/com/redhat/cloud/notifications/routers/UserConfigService.java:78]: Merging should be avoided indeed because it runs things concurrently, just like `Uni.combine().all()`. `concatenate` on the other hand runs things sequentially which should be the preferred way to interact with the database.

**Comment by gwenneg** [comment]
The latest build finally passed, but there are still a few things still in progress, including:
- `LifecycleITest` is unstable but that is also true with `master`, at least on my machine
- I had to disable a call in `EndpointServiceTest.testEndpointValidation` because of a bug which is probably coming from Quarkus or one of its dependencies
- Micrometer seems to be broken for now because of the removal of the JDBC datasource, I have to investigate this (not started)

The old code didn't include any kind of transactional behavior and that is still true with the new one. That's intentional even if I think we'll need to introduce transactions shortly. This could be done in this PR or not. My goal was to migrate from R2DBC to Hibernate Reactive with as few code changes as possible to make the PR easier to review.

We had a discussion with @pilhuhn today about the DTOs vs no-DTOs approach. Both have their advantages and drawbacks. While I have a strong opinion about the need to DTOs for the sake of maintainability, DTOs come at a cost in terms of application performance (memory consumption, pressure on GC...) and that's equally important to me. I'll give the no-DTOs approach a try on a parallel branch derived from this one and see how it goes.

File: src/main/resources/application.properties [hunk 4]
```diff
@@ -14,11 +14,13 @@ mp.messaging.incoming.ingress.value.deserializer=org.apache.kafka.common.seriali
 quarkus.datasource.db-kind=postgresql
 quarkus.datasource.username=postgres
 quarkus.datasource.password=postgres
-quarkus.datasource.jdbc.url=jdbc:postgresql://127.0.0.1:5432/notifications
-quarkus.datasource.jdbc.max-size=4
+quarkus.datasource.jdbc=false
+quarkus.datasource.reactive.url=postgresql://127.0.0.1:5432/notifications
+quarkus.datasource.reactive.max-size=4
```
Comment (pilhuhn) [inline src/main/resources/application.properties:19]: I think we can allow for some more connections. The jdbc setting was to limit the flyway ones, but r2dbc did not have that setting.

Comment (gwenneg) [reply src/main/resources/application.properties:19]: I fixed that on the other branch.

**Comment by gwenneg** [comment]
Superseded by #148.
